mtp_cashbook/apps/cashbook/tests/test_views.py

```diff
@@ -232,6 +233,127 @@ def test_new_credits_submit(self, mock_create_transaction):
+            )
+            # get manual credits
+            rsps.add(
+                rsps.GET,
+                api_url('/credits/?resolution=manual&status=credit_pending&offset=0&limit=100&ordering=-received_at'),
+                json=wrap_response_data(),
+                status=200,
+                match_querystring=True,
+            )
+            # create batch
+            rsps.add(
+                rsps.POST,
+                api_url('/credits/batches/'),
+                status=201,
+            )
+            rsps.add(
+                rsps.POST,
+                api_url('/credits/actions/credit/'),
+                status=204,
+            )
+            rsps.add(
+                rsps.POST,
+                api_url('/credits/actions/credit/'),
+                status=204,
+            )
+            # REDIRECT after success
+            # get active batches
+            rsps.add(
+                rsps.GET,
+                api_url('/credits/batches/'),
+                json=wrap_response_data(PROCESSING_BATCH),
+                status=200,
+            )
+            # get incomplete credits
+            rsps.add(
+                rsps.GET,
+                api_url('/credits/?resolution=pending&pk=1&pk=2'),
+                json=wrap_response_data(),
+                status=200,
+                match_querystring=True,
+            )
+            # get complete credits
+            rsps.add(
+                rsps.GET,
+                api_url('/credits/?resolution=credited&pk=1&pk=2'),
+                json=wrap_response_data(CREDIT_1, CREDIT_2),
+                status=200,
+                match_querystring=True,
+            )
+            # delete completed batch
+            rsps.add(
+                rsps.DELETE,
+                api_url('/credits/batches/%s/' % PROCESSING_BATCH['id']),
+                status=200,
+            )
+            # get new credits
+            rsps.add(
+                rsps.GET,
+                api_url('/credits/?ordering=-received_at&offset=0&limit=100&status=credit_pending&resolution=pending'),
+                json=wrap_response_data(),
+                status=200,
+                match_querystring=True,
+            )
+            # get manual credits
+            rsps.add(
+                rsps.GET,
+                api_url('/credits/?resolution=manual&status=credit_pending&offset=0&limit=100&ordering=-received_at'),
+                json=wrap_response_data(),
+                status=200,
+                match_querystring=True,
+            )
+
+            self.login()
+            response = self.client.post(
+                self.url,
+                data={'credits': [1, 2], 'submit_new': None},
+                follow=True
+            )
+
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, '2 credits sent to NOMIS')
+        logger_error_messages = [call[0][0] for call in mock_logger.error.call_args_list]
```

Comment:
I really don't like asserting against calls to the logging framework. Logging is ancillary functionality, there should not be a dependence on the veracity of the code flow based on what the logger sees. I would need to look at the code a bit closer but my hunch is that this functionality can be replicated by asserting against a spy function,

Reply:
the purpose of this feature is to log an error, so really that _is_ the thing we want to assert is happening

Reply:
I guess that's true in this case, since there's a functional requirement for the logging message to be added (as opposed to non-functional) we need to test that requirement.